flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -212,31 +269,46 @@ private void flushCurrentSortBuffer() throws IOException {
         if (currentSortBuffer.hasRemaining()) {
             fileWriter.startNewRegion();
 
+            List<BufferWithChannel> toWrite = new ArrayList<>();
+            Queue<MemorySegment> segments = getWriteBuffers();
+
             while (currentSortBuffer.hasRemaining()) {
-                BufferWithChannel bufferWithChannel =
-                        currentSortBuffer.copyIntoSegment(writeBuffer);
-                Buffer buffer = bufferWithChannel.getBuffer();
-                int subpartitionIndex = bufferWithChannel.getChannelIndex();
+                if (segments.isEmpty()) {
+                    fileWriter.writeBuffers(toWrite);
+                    toWrite.clear();
+                    segments = getWriteBuffers();
+                }
 
-                writeCompressedBufferIfPossible(buffer, subpartitionIndex);
+                BufferWithChannel bufferWithChannel =
+                        currentSortBuffer.copyIntoSegment(checkNotNull(segments.poll()));
+                toWrite.add(compressBufferIfPossible(bufferWithChannel));
             }
+
+            fileWriter.writeBuffers(toWrite);
         }
 
         currentSortBuffer.release();
     }
 
-    private void writeCompressedBufferIfPossible(Buffer buffer, int targetSubpartition)
-            throws IOException {
-        updateStatistics(buffer, targetSubpartition);
+    private Queue<MemorySegment> getWriteBuffers() {
+        synchronized (lock) {
+            checkState(!writeBuffers.isEmpty(), "Task has been canceled.");
+            return new ArrayDeque<>(writeBuffers);
+        }
+    }
 
-        try {
-            if (canBeCompressed(buffer)) {
-                buffer = bufferCompressor.compressToIntermediateBuffer(buffer);
-            }
-            fileWriter.writeBuffer(buffer, targetSubpartition);
-        } finally {
-            buffer.recycleBuffer();
+    private BufferWithChannel compressBufferIfPossible(BufferWithChannel bufferWithChannel) {
+        Buffer buffer = bufferWithChannel.getBuffer();
+        int channelIndex = bufferWithChannel.getChannelIndex();
+
+        updateStatistics(buffer, channelIndex);
```

Comment:
After the modification, `updateStatistics` method called here seems to be a bit inconsistent with the name of this method. We might move the statistics out.